homeassistant/components/weather/__init__.py

```diff
@@ -173,23 +451,69 @@ def ozone(self) -> float | None:
 
     @property
     def visibility(self) -> float | None:
-        """Return the visibility in native units."""
+        """Return the visibility for backward compatibility.
+
+        Should not be set by integrations.
+        """
         return self._attr_visibility
 
     @property
-    def visibility_unit(self) -> str | None:
+    def native_visibility(self) -> float | None:
+        """Return the visibility in native units."""
+        return self._attr_native_visibility
+
+    @property
+    def native_visibility_unit(self) -> str | None:
         """Return the native unit of measurement for visibility."""
-        return self._attr_visibility_unit
+        return self._attr_native_visibility_unit
+
+    @property
+    def visibility_unit(self) -> str:
+        """Return the converted unit of measurement for visibility.
+
+        Should not be set by integrations.
+        """
+        if (
+            weather_option_visibility_uom := self._weather_option_visibility_uom
+        ) is not None:
+            return weather_option_visibility_uom
+
+        if (visibility_unit := self._attr_visibility_unit) is not None:
+            return visibility_unit
+
+        if (native_visibility_unit := self.native_visibility_unit) is not None:
+            return native_visibility_unit
+
+        return self.hass.config.units.length_unit
 
     @property
     def forecast(self) -> list[Forecast] | None:
         """Return the forecast in native units."""
         return self._attr_forecast
 
     @property
-    def precipitation_unit(self) -> str | None:
+    def native_precipitation_unit(self) -> str | None:
         """Return the native unit of measurement for accumulated precipitation."""
-        return self._attr_precipitation_unit
+        return self._attr_native_precipitation_unit
+
+    @property
+    def precipitation_unit(self) -> str:
```

Comment:
Do like this for backwards compatibility:
```suggestion
        """Return the precipitation unit for backward compatibility.
        Should not be set by integrations.
        """
        return self._attr_precipitation_unit

    @property
    def _precipitation_unit(self) -> str:
```